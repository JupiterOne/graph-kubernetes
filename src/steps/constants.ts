import {
  RelationshipClass,
  RelationshipDirection,
  StepMappedRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const CLUSTER_ENTITY_DATA_KEY = 'entity:cluster';

export enum IntegrationSteps {
  NETWORK_POLICIES = 'fetch-network-policies',
  SERVICE_ACCOUNTS = 'fetch-service-accounts',
  USERS = 'fetch-users',
  ROLES = 'fetch-roles',
  CLUSTER_ROLES = 'fetch-cluster-roles',
  ROLE_BINDINGS = 'fetch-role-bindings',
  CLUSTER_ROLE_BINDINGS = 'fetch-cluster-role-bindings',
  CLUSTERS = 'fetch-clusters',
  BUILD_CLUSTER_RESOURCES_RELATIONSHIPS = 'build-cluster-resources-relationships',
  BUILD_CLUSTER_AKS_RELATIONSHIPS = 'build-cluster-aks-relationships',
  BUILD_CLUSTER_GKE_RELATIONSHIPS = 'build-cluster-gke-relationships',
  NAMESPACES = 'fetch-namespaces',
  NODES = 'fetch-nodes',
  SERVICES = 'fetch-services',
  DEPLOYMENTS = 'fetch-deployments',
  BUILD_CONTAINER_SPEC_VOLUME_RELATIONSHIPS = 'build-container-spec-volume-relationships',
  REPLICASETS = 'fetch-replica-sets',
  STATEFULSETS = 'fetch-stateful-sets',
  DAEMONSETS = 'fetch-daemon-sets',
  JOBS = 'fetch-jobs',
  CRONJOBS = 'fetch-cron-jobs',
  CONFIGMAPS = 'fetch-config-maps',
  SECRETS = 'fetch-secrets',
  IMAGES = 'fetch-images',
  PODS = 'fetch-pods',
}

export const Entities = {
  NETWORK_POLICY: {
    _type: 'kube_network_policy',
    _class: ['Configuration'],
    resourceName: 'Kubernetes Network Policy',
  },
  SERVICE_ACCOUNT: {
    _type: 'kube_service_account',
    _class: ['User'],
    resourceName: 'Kubernetes Service Account',
  },
  USER: {
    _type: 'kube_user',
    _class: ['User'],
    resourceName: 'Kubernetes User',
  },
  ROLE: {
    _type: 'kube_role',
    _class: ['AccessRole'],
    resourceName: 'Kubernetes Role',
  },
  CLUSTER_ROLE: {
    _type: 'kube_cluster_role',
    _class: ['AccessRole'],
    resourceName: 'Kubernetes Cluster Role',
  },
  ROLE_BINDING: {
    _type: 'kube_role_binding',
    _class: ['AccessPolicy'],
    resourceName: 'Kubernetes Role Binding',
  },
  CLUSTER_ROLE_BINDING: {
    _type: 'kube_cluster_role_binding',
    _class: ['AccessPolicy'],
    resourceName: 'Kubernetes Cluster Role Binding',
  },
  CLUSTER: {
    _type: 'kube_cluster',
    _class: ['Cluster'],
    resourceName: 'Kubernetes Cluster',
  },
  AZURE_KUBERNETES_CLUSTER: {
    _type: 'azure_kubernetes_cluster',
    _class: ['Cluster'],
    resourceName: 'Azure Kubernetes Cluster',
  },
  GOOGLE_KUBERNETES_CLUSTER: {
    _type: 'google_container_cluster',
    _class: ['Cluster'],
    resourceName: 'Google Kubernetes Cluster',
  },
  NAMESPACE: {
    _type: 'kube_namespace',
    _class: ['Group'],
    resourceName: 'Kubernetes Namespace',
  },
  DEPLOYMENT: {
    _type: 'kube_deployment',
    _class: ['Deployment'],
    resourceName: 'Kubernetes Deployment',
  },
  CONTAINER_SPEC: {
    _type: 'kube_container_spec',
    _class: ['Configuration'],
    resourceName: 'Kubernetes Container Spec',
  },
  VOLUME: {
    _type: 'kube_volume',
    _class: ['Disk'],
    resourceName: 'Kubernetes Volume',
  },
  REPLICASET: {
    _type: 'kube_replica_set',
    _class: ['Deployment'],
    resourceName: 'Kubernetes ReplicaSet',
  },
  STATEFULSET: {
    _type: 'kube_stateful_set',
    _class: ['Deployment'],
    resourceName: 'Kubernetes StatefulSet',
  },
  DAEMONSET: {
    _type: 'kube_daemon_set',
    _class: ['Deployment'],
    resourceName: 'Kubernetes DaemonSet',
  },
  NODE: {
    _type: 'kube_node',
    _class: ['Host'],
    resourceName: 'Kubernetes Node',
  },
  JOB: {
    _type: 'kube_job',
    _class: ['Task'],
    resourceName: 'Kubernetes Job',
  },
  CRONJOB: {
    _type: 'kube_cron_job',
    _class: ['Task'],
    resourceName: 'Kubernetes CronJob',
  },
  CONFIGMAP: {
    _type: 'kube_config_map',
    _class: ['Configuration'],
    resourceName: 'Kubernetes ConfigMap',
  },
  SECRET: {
    _type: 'kube_secret',
    _class: ['Vault'],
    resourceName: 'Kubernetes Secret',
  },
  SERVICE: {
    _type: 'kube_service',
    _class: ['Service'],
    resourceName: 'Kubernetes Service',
  },
  IMAGE: {
    _type: 'kube_image',
    _class: ['Image'],
    resourceName: 'Kubernetes Image',
  },
  POD: {
    _type: 'kube_pod',
    _class: ['Task'],
    resourceName: 'Kubernetes Pod',
  },
  CONTAINER: {
    _type: 'kube_container',
    _class: ['Container'],
    resourceName: 'Kubernetes Container',
  },
};

export const Relationships = {
  CLUSTER_CONTAINS_NAMESPACE: {
    _type: 'kube_cluster_contains_namespace',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.NAMESPACE._type,
  },
  CLUSTER_CONTAINS_CLUSTER_ROLE: {
    _type: 'kube_cluster_contains_cluster_role',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.CLUSTER_ROLE._type,
  },
  CLUSTER_CONTAINS_CLUSTER_ROLE_BINDING: {
    _type: 'kube_cluster_contains_cluster_role_binding',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.CLUSTER_ROLE_BINDING._type,
  },
  NAMESPACE_CONTAINS_SERVICE: {
    _type: 'kube_namespace_contains_service',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.SERVICE._type,
  },
  NAMESPACE_CONTAINS_DEPLOYMENT: {
    _type: 'kube_namespace_contains_deployment',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.DEPLOYMENT._type,
  },
  NAMESPACE_CONTAINS_REPLICASET: {
    _type: 'kube_namespace_contains_replica_set',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.REPLICASET._type,
  },
  NAMESPACE_CONTAINS_STATEFULSET: {
    _type: 'kube_namespace_contains_stateful_set',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.STATEFULSET._type,
  },
  NAMESPACE_CONTAINS_DAEMONSET: {
    _type: 'kube_namespace_contains_daemon_set',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.DAEMONSET._type,
  },
  NAMESPACE_CONTAINS_CRONJOB: {
    _type: 'kube_namespace_contains_cron_job',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.CRONJOB._type,
  },
  NAMESPACE_CONTAINS_JOB: {
    _type: 'kube_namespace_contains_job',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.JOB._type,
  },
  NAMESPACE_CONTAINS_CONFIGMAP: {
    _type: 'kube_namespace_contains_config_map',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.CONFIGMAP._type,
  },
  NAMESPACE_CONTAINS_SECRET: {
    _type: 'kube_namespace_contains_secret',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.SECRET._type,
  },
  NAMESPACE_CONTAINS_NETWORK_POLICY: {
    _type: 'kube_namespace_contains_network_policy',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.NETWORK_POLICY._type,
  },
  NAMESPACE_CONTAINS_SERVICE_ACCOUNT: {
    _type: 'kube_namespace_contains_service_account',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.SERVICE_ACCOUNT._type,
  },
  NAMESPACE_CONTAINS_ROLE: {
    _type: 'kube_namespace_contains_role',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.ROLE._type,
  },
  NAMESPACE_CONTAINS_ROLE_BINDING: {
    _type: 'kube_namespace_contains_role_binding',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.ROLE_BINDING._type,
  },
  DEPLOYMENT_MANAGES_REPLICASET: {
    _type: 'kube_deployment_manages_replica_set',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.DEPLOYMENT._type,
    targetType: Entities.REPLICASET._type,
  },
  DEPLOYMENT_USES_CONTAINER_SPEC: {
    _type: 'kube_deployment_uses_container_spec',
    _class: RelationshipClass.USES,
    sourceType: Entities.DEPLOYMENT._type,
    targetType: Entities.CONTAINER_SPEC._type,
  },
  CONTAINER_SPEC_USES_VOLUME: {
    _type: 'kube_container_spec_uses_volume',
    _class: RelationshipClass.USES,
    sourceType: Entities.CONTAINER_SPEC._type,
    targetType: Entities.VOLUME._type,
  },
  CRONJOB_MANAGES_JOB: {
    _type: 'kube_cron_job_manages_job',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.CRONJOB._type,
    targetType: Entities.JOB._type,
  },
  NODE_HAS_IMAGE: {
    _type: 'kube_node_has_image',
    _class: RelationshipClass.HAS,
    sourceType: Entities.NODE._type,
    targetType: Entities.IMAGE._type,
  },
  REPLICA_SET_USES_IMAGE: {
    _type: 'kube_replica_set_uses_image',
    _class: RelationshipClass.USES,
    sourceType: Entities.REPLICASET._type,
    targetType: Entities.IMAGE._type,
  },
  REPLICASET_MANAGES_POD: {
    _type: 'kube_replica_set_manages_pod',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.REPLICASET._type,
    targetType: Entities.POD._type,
  },
  STATEFULSET_MANAGES_POD: {
    _type: 'kube_stateful_set_manages_pod',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.STATEFULSET._type,
    targetType: Entities.POD._type,
  },
  JOB_MANAGES_POD: {
    _type: 'kube_job_manages_pod',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.JOB._type,
    targetType: Entities.POD._type,
  },
  NODE_HAS_POD: {
    _type: 'kube_node_has_pod',
    _class: RelationshipClass.HAS,
    sourceType: Entities.NODE._type,
    targetType: Entities.POD._type,
  },
  POD_CONTAINS_CONTAINER: {
    _type: 'kube_pod_contains_container',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.POD._type,
    targetType: Entities.CONTAINER._type,
  },
};

export const MappedRelationships: Record<
  string,
  StepMappedRelationshipMetadata
> = {
  CLUSTER_IS_AKS_CLUSTER: {
    _type: 'kube_cluster_is_azure_kubernetes_cluster',
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.AZURE_KUBERNETES_CLUSTER._type,
    _class: RelationshipClass.IS,
    direction: RelationshipDirection.FORWARD,
  },
  CLUSTER_IS_GKE_CLUSTER: {
    _type: 'kube_cluster_is_google_container_cluster',
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.GOOGLE_KUBERNETES_CLUSTER._type,
    _class: RelationshipClass.IS,
    direction: RelationshipDirection.FORWARD,
  },
};
