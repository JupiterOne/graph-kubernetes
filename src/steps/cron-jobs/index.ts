import {
  createDirectRelationship,
  getRawData,
  IntegrationMissingKeyError,
  IntegrationStep,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { V1CronJob } from '@kubernetes/client-node';
import { IntegrationConfig, IntegrationStepContext } from '../../config';
import { Entities, IntegrationSteps, Relationships } from '../constants';
import { createCronJobEntity } from './converters';
import getOrCreateAPIClient from '../../kubernetes/getOrCreateAPIClient';
import {
  createContainerSpecEntity,
  getContainerSpecKey,
} from '../deployments/converters';

export async function fetchCronJobs(
  context: IntegrationStepContext,
): Promise<void> {
  const { instance, jobState } = context;
  const { config } = instance;

  const client = getOrCreateAPIClient(config);

  await jobState.iterateEntities(
    {
      _type: Entities.NAMESPACE._type,
    },
    async (namespaceEntity) => {
      await client.iterateNamespacedCronJobs(
        namespaceEntity.name as string,
        async (cronJob) => {
          const cronJobEntity = createCronJobEntity(cronJob);
          await jobState.addEntity(cronJobEntity);

          await jobState.addRelationship(
            createDirectRelationship({
              _class: RelationshipClass.CONTAINS,
              from: namespaceEntity,
              to: cronJobEntity,
            }),
          );

          for (const container of cronJob.spec?.jobTemplate?.spec?.template
            ?.spec?.containers ||
            [] ||
            []) {
            const cronJobContainerspecEntity = createContainerSpecEntity(
              namespaceEntity.name as string,
              container,
            );
            // Check if the entity is already present in jobState
            if (jobState.hasKey(cronJobContainerspecEntity._key)) {
              continue; // Skip to the next iteration if the entity is already present
            }

            await jobState.addEntity(cronJobContainerspecEntity);
          }
        },
      );
    },
  );
}

export async function buildContainerSpecCronJobRelationship(
  context: IntegrationStepContext,
): Promise<void> {
  const { jobState } = context;

  await jobState.iterateEntities(
    {
      _type: Entities.NAMESPACE._type,
    },
    async (namespaceEntity) => {
      // Iterate over cron job entities within the current namespace
      await jobState.iterateEntities(
        {
          _type: Entities.CRONJOB._type,
        },
        async (cronJobEntity) => {
          const rawNode = getRawData<V1CronJob>(cronJobEntity);
          const cronJobContainers =
            rawNode?.spec?.jobTemplate?.spec?.template?.spec?.containers;
          if (cronJobContainers) {
            for (const container of cronJobContainers) {
              const containerSpecKey = getContainerSpecKey(
                namespaceEntity.name as string,
                container.name,
              );

              if (!containerSpecKey) {
                throw new IntegrationMissingKeyError(
                  `Cannot build Relationship.
                  Error: Missing Key.
                  containerSpecKey: ${containerSpecKey}`,
                );
              }

              await jobState.addRelationship(
                createDirectRelationship({
                  _class: RelationshipClass.USES,
                  toKey: containerSpecKey,
                  toType: Entities.CONTAINER_SPEC._type,
                  fromKey: cronJobEntity._key,
                  fromType: Entities.CRONJOB._type,
                }),
              );
            }
          }
        },
      );
    },
  );
}

export const cronJobsSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.CRONJOBS,
    name: 'Fetch CronJobs',
    entities: [Entities.CRONJOB],
    relationships: [Relationships.NAMESPACE_CONTAINS_CRONJOB],
    dependsOn: [IntegrationSteps.NAMESPACES],
    executionHandler: fetchCronJobs,
  },
  {
    id: IntegrationSteps.CONTAINER_SPEC_HAS_CRON_JOB,
    name: 'Build Container Spec HAS CronJob relationship',
    entities: [],
    relationships: [Relationships.CRONJOB_USES_CONTAINER_SPEC],
    dependsOn: [IntegrationSteps.CRONJOBS],
    executionHandler: buildContainerSpecCronJobRelationship,
  },
];
