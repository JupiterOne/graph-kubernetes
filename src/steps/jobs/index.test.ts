import { buildContainerSpecJobRelationship, fetchJobs } from '.';
import { fetchCronJobs } from '../cron-jobs';
import { fetchNamespaces } from '../namespaces';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities, Relationships } from '../constants';
import { RelationshipClass } from '@jupiterone/data-model';

describe('#fetchJobs1', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'fetchJobs',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [
        fetchNamespaces,
        fetchCronJobs,
        fetchJobs,
        buildContainerSpecJobRelationship,
      ],
      entitySchemaMatchers: [
        {
          _type: Entities.JOB._type,
          matcher: {
            _class: ['Task'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'kube_job' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                displayName: { type: 'string' },
                generation: { type: 'number' },
                deletionGracePeriodSeconds: { type: 'string' },
                resourceVersion: { type: 'string' },
                createdOn: { type: 'number' },
                deletedOn: { type: 'number' },

                activeDeadlineSeconds: { type: 'number' },
                backoffLimit: { type: 'number' },
                completions: { type: 'number' },
                manualSelector: { type: 'boolean' },
                parallelism: { type: 'number' },
                ttlSecondsAfterFinished: { type: 'number' },
                'status.active': { type: 'number' },
                'status.startTime': { type: 'number' },
                'status.completionTime': { type: 'number' },
                'status.failed': { type: 'number' },
                'status.succeeded': { type: 'number' },
              },
            },
          },
        },
      ],
      relationshipSchemaMatchers: [
        {
          _type: Relationships.NAMESPACE_CONTAINS_JOB._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.CONTAINS },
                _type: { const: 'kube_namespace_contains_job' },
              },
            },
          },
        },
        {
          _type: Relationships.CRONJOB_MANAGES_JOB._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.MANAGES },
                _type: { const: 'kube_cron_job_manages_job' },
              },
            },
          },
        },
        {
          _type: Relationships.CONTAINER_SPEC_HAS_JOB._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.HAS },
                _type: { const: 'kube_container_spec_has_job' },
              },
            },
          },
        },
      ],
    });
  }, 500000);
});
