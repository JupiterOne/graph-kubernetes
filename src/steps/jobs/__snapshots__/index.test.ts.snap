// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchJobs1 should collect data: jobState 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Group",
      ],
      "_key": "0db73bce-0801-4b25-867a-a9fe958fa96d",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-21T07:44:54.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "default",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-21T07:44:54.000Z,
                },
              ],
              "name": "default",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "42",
              "selfLink": undefined,
              "uid": "0db73bce-0801-4b25-867a-a9fe958fa96d",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718955894000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "default",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "default",
      "resourceVersion": "42",
      "status.phase": "Active",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "a3b7fbce-2d26-47e5-97b6-3ccc51b40932",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-21T07:44:54.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "kube-node-lease",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-21T07:44:54.000Z,
                },
              ],
              "name": "kube-node-lease",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "39",
              "selfLink": undefined,
              "uid": "a3b7fbce-2d26-47e5-97b6-3ccc51b40932",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718955894000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-node-lease",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-node-lease",
      "resourceVersion": "39",
      "status.phase": "Active",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "4055ce72-f3cb-457d-8003-c10a01017dd8",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-21T07:44:54.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "kube-public",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-21T07:44:54.000Z,
                },
              ],
              "name": "kube-public",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "31",
              "selfLink": undefined,
              "uid": "4055ce72-f3cb-457d-8003-c10a01017dd8",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718955894000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-public",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-public",
      "resourceVersion": "31",
      "status.phase": "Active",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "dae8939b-da22-491d-a98b-c8fdf0e2f119",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-21T07:44:54.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "kube-system",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-21T07:44:54.000Z,
                },
              ],
              "name": "kube-system",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "12",
              "selfLink": undefined,
              "uid": "dae8939b-da22-491d-a98b-c8fdf0e2f119",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718955894000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-system",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-system",
      "resourceVersion": "12",
      "status.phase": "Active",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "c34b14f9-fe12-4a1b-9b58-894760f6a770",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": {
                "kubectl.kubernetes.io/last-applied-configuration": "{"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{},"labels":{"addonmanager.kubernetes.io/mode":"Reconcile","kubernetes.io/minikube-addons":"dashboard"},"name":"kubernetes-dashboard"}}
",
              },
              "creationTimestamp": 2024-06-21T07:45:06.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "addonmanager.kubernetes.io/mode": "Reconcile",
                "kubernetes.io/metadata.name": "kubernetes-dashboard",
                "kubernetes.io/minikube-addons": "dashboard",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:annotations": {
                        ".": {},
                        "f:kubectl.kubernetes.io/last-applied-configuration": {},
                      },
                      "f:labels": {
                        ".": {},
                        "f:addonmanager.kubernetes.io/mode": {},
                        "f:kubernetes.io/metadata.name": {},
                        "f:kubernetes.io/minikube-addons": {},
                      },
                    },
                  },
                  "manager": "kubectl-client-side-apply",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-21T07:45:06.000Z,
                },
              ],
              "name": "kubernetes-dashboard",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "315",
              "selfLink": undefined,
              "uid": "c34b14f9-fe12-4a1b-9b58-894760f6a770",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718955906000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kubernetes-dashboard",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kubernetes-dashboard",
      "resourceVersion": "315",
      "status.phase": "Active",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "kube_namespace",
  ],
  "numCollectedEntities": 5,
  "numCollectedRelationships": 0,
}
`;
